File: php/tests/backup_integrity_test.php
<?php
/**
 * backup_integrity_test.php
 *
 * Unique backup names + abort incomplete backup (Phase 2).
 *
 * Usage:
 *   php connectors/php/tests/backup_integrity_test.php
 */

function fail(string $msg): void {
    fwrite(STDERR, "FAIL: {$msg}\n");
    exit(1);
}

function assert_true($cond, string $msg): void {
    if (!$cond) {
        fail($msg);
    }
    echo "  OK  {$msg}\n";
}

$managerPath = realpath(__DIR__ . '/../backup_manager.php');
if (!$managerPath || !file_exists($managerPath)) {
    fail('backup_manager.php not found');
}
require_once $managerPath;

$tmp = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'patcherly-php-bk-' . bin2hex(random_bytes(4));
$backupRoot = $tmp . DIRECTORY_SEPARATOR . 'backups';
$targetRoot = $tmp . DIRECTORY_SEPARATOR . 'target';
if (!mkdir($backupRoot, 0700, true) || !mkdir($targetRoot, 0700, true)) {
    fail('could not create temp dirs');
}

putenv('PATCHERLY_TARGET_ROOTS=' . $targetRoot);
$prevCwd = getcwd();
chdir($targetRoot);

$bm = new AgentBackupManager($backupRoot);

// ---- 1. Same-basename collision + restore round-trip ----
$dirA = $targetRoot . DIRECTORY_SEPARATOR . 'theme-a';
$dirB = $targetRoot . DIRECTORY_SEPARATOR . 'theme-b';
mkdir($dirA, 0700, true);
mkdir($dirB, 0700, true);
$fileA = $dirA . DIRECTORY_SEPARATOR . 'functions.php';
$fileB = $dirB . DIRECTORY_SEPARATOR . 'functions.php';
file_put_contents($fileA, "content-A\n");
file_put_contents($fileB, "content-B\n");

$meta = $bm->createBackup('collision', [$fileA, $fileB], false, true);
$leaves = [];
foreach ($meta['manifest'] as $info) {
    $leaves[] = basename($info['backup_path']);
}
assert_true(count($leaves) === 2, 'two files backed up');
assert_true($leaves[0] !== $leaves[1], 'backup leaf names must differ for same basename');

file_put_contents($fileA, "MUTATED-A\n");
file_put_contents($fileB, "MUTATED-B\n");
$ok = $bm->restoreBackup($meta['backup_dir']);
assert_true($ok === true, 'restore succeeded');
assert_true(file_get_contents($fileA) === "content-A\n", 'file A restored');
assert_true(file_get_contents($fileB) === "content-B\n", 'file B restored');

// ---- 2. Outside allowed roots → abort ----
$outsideRoot = dirname($targetRoot) . DIRECTORY_SEPARATOR . 'outside';
mkdir($outsideRoot, 0700, true);
$outside = $outsideRoot . DIRECTORY_SEPARATOR . 'evil.txt';
file_put_contents($outside, "evil\n");
$threw = false;
try {
    $bm->createBackup('outside', [$fileA, $outside], false, false);
} catch (Exception $e) {
    $threw = true;
    assert_true(
        strpos($e->getMessage(), 'outside allowed target roots') !== false,
        'outside-roots error message'
    );
}
assert_true($threw, 'createBackup throws for path outside allowed roots');

// ---- 3. Missing file skip-OK ----
$existing = $targetRoot . DIRECTORY_SEPARATOR . 'exists.txt';
$missing = $targetRoot . DIRECTORY_SEPARATOR . 'will-create.txt';
file_put_contents($existing, "exists\n");
$meta2 = $bm->createBackup('missing-ok', [$existing, $missing], false, false);
assert_true(count($meta2['files']) === 1, 'missing file skipped; one file backed up');

chdir($prevCwd);

// Cleanup best-effort
function rrmdir($dir) {
    if (!is_dir($dir)) return;
    foreach (scandir($dir) ?: [] as $item) {
        if ($item === '.' || $item === '..') continue;
        $path = $dir . DIRECTORY_SEPARATOR . $item;
        is_dir($path) ? rrmdir($path) : @unlink($path);
    }
    @rmdir($dir);
}
rrmdir($tmp);

echo "All backup integrity tests passed.\n";
exit(0);

File: python/tests/backup_integrity_test.py
#!/usr/bin/env python3
"""
backup_integrity_test.py

Unique backup names + abort incomplete backup (Phase 2 production readiness).

1. Same-basename two-file collision: backup + restore round-trip keeps both.
2. Existing file snapshot failure → create_backup aborts (raises).
3. Missing listed file is skip-OK.

Run:  python connectors/python/tests/backup_integrity_test.py
"""

from __future__ import annotations

import asyncio
import os
import sys
import tempfile
import unittest
from pathlib import Path
from unittest import mock

_CONNECTOR_DIR = Path(__file__).resolve().parent.parent
if str(_CONNECTOR_DIR) not in sys.path:
    sys.path.insert(0, str(_CONNECTOR_DIR))

from backup_manager import AgentBackupManager  # noqa: E402


class BackupIntegrityTest(unittest.TestCase):
    def setUp(self) -> None:
        self.tmp = tempfile.TemporaryDirectory()
        self.root = Path(self.tmp.name).resolve()
        self.backup_root = self.root / "backups"
        self.target_root = self.root / "target"
        self.backup_root.mkdir()
        self.target_root.mkdir()
        self.bm = AgentBackupManager(str(self.backup_root))

    def tearDown(self) -> None:
        self.tmp.cleanup()

    def test_same_basename_collision_round_trip(self) -> None:
        dir_a = self.target_root / "theme-a"
        dir_b = self.target_root / "theme-b"
        dir_a.mkdir()
        dir_b.mkdir()
        file_a = dir_a / "functions.php"
        file_b = dir_b / "functions.php"
        file_a.write_text("content-A\n", encoding="utf-8")
        file_b.write_text("content-B\n", encoding="utf-8")

        meta = asyncio.run(
            self.bm.create_backup("collision", [str(file_a), str(file_b)], compress=False, verify=True)
        )

        leaf_names = [Path(info["backup_path"]).name for info in meta.manifest.values()]
        self.assertEqual(len(leaf_names), 2)
        self.assertNotEqual(leaf_names[0], leaf_names[1], "backup leaf names must differ")
        self.assertTrue(all("functions.php" in n for n in leaf_names))

        file_a.write_text("MUTATED-A\n", encoding="utf-8")
        file_b.write_text("MUTATED-B\n", encoding="utf-8")
        ok = asyncio.run(self.bm.restore_backup(meta.backup_dir))
        self.assertTrue(ok)
        self.assertEqual(file_a.read_text(encoding="utf-8"), "content-A\n")
        self.assertEqual(file_b.read_text(encoding="utf-8"), "content-B\n")

    def test_abort_when_existing_file_snapshot_fails(self) -> None:
        good = self.target_root / "good.txt"
        bad = self.target_root / "bad.txt"
        good.write_text("good\n", encoding="utf-8")
        bad.write_text("bad\n", encoding="utf-8")

        async def boom(path: Path):
            if path.resolve() == bad.resolve():
                raise OSError("simulated read failure")
            with open(path, "rb") as f:
                return f.read()

        with mock.patch.object(self.bm, "_read_file_async", side_effect=boom):
            with self.assertRaises(ValueError) as ctx:
                asyncio.run(
                    self.bm.create_backup("fail-snap", [str(good), str(bad)], compress=False, verify=False)
                )
        self.assertIn("Failed to backup existing file", str(ctx.exception))

    def test_missing_file_is_skip_ok(self) -> None:
        existing = self.target_root / "exists.txt"
        missing = self.target_root / "will-create.txt"
        existing.write_text("exists\n", encoding="utf-8")

        meta = asyncio.run(
            self.bm.create_backup(
                "missing-ok", [str(existing), str(missing)], compress=False, verify=False
            )
        )
        self.assertEqual(len(meta.files), 1)
        self.assertEqual(Path(meta.files[0]).resolve(), existing.resolve())

    def test_unique_backup_file_name_helper(self) -> None:
        a = self.target_root / "a" / "foo.txt"
        b = self.target_root / "b" / "foo.txt"
        name_a = self.bm._unique_backup_file_name(a.resolve())
        name_b = self.bm._unique_backup_file_name(b.resolve())
        self.assertNotEqual(name_a, name_b)
        self.assertTrue(name_a.endswith("foo.txt") or "foo.txt" in name_a)
        self.assertNotIn(os.sep, name_a)
        self.assertNotIn("/", name_a)


if __name__ == "__main__":
    unittest.main()
